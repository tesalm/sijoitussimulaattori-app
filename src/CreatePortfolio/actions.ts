import { Dispatch } from 'redux';

export enum ActionType {
  CreateNewPortfolio = '[CreatingPortfolio] Create new portfolio',
}

export type CreatePortfolioAction = CreateNewPortfolio;

interface CreateNewPortfolio {
  type: ActionType.CreateNewPortfolio;
  name: string;
  amount: number;
}

const createNewPortfolioAction = (
  name: string,
  amount: number
): CreateNewPortfolio => ({
  type: ActionType.CreateNewPortfolio,
  name: name,
  amount: amount,
});

const CreateNewPortfolio = (name: string, amount: number) => async (
  dispatch: Dispatch<CreateNewPortfolio>
) => dispatch(createNewPortfolioAction(name, amount));

export { CreateNewPortfolio };
