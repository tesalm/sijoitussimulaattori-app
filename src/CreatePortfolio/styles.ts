import { StyleSheet } from 'react-native';

import { Colors } from '../App/colors';

const CreatePortfolioStyles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'stretch',
    marginTop: 8,
    marginBottom: 8,
    marginLeft: 16,
    marginRight: 16,
  },
  main: {
    fontFamily: 'Roboto',
    fontSize: 16,
    fontWeight: 'bold',
    textAlign: 'left',
    color: Colors.baseColor,
  },
  textinput: {
    marginTop: 8,
    marginBottom: 8,
    borderBottomWidth: 1,
    borderBottomColor: Colors.textInputBottomBorder,
    height: 40,
    color: Colors.textInputText,
  },
  error: {
    color: Colors.errorText,
  },
});

export { CreatePortfolioStyles };
