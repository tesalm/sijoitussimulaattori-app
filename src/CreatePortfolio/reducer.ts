import { ActionType, CreatePortfolioAction } from './actions';

export interface CreatingPortfolio {
  amount: number;
  name: string;
}

const initialState: CreatingPortfolio = {
  amount: NaN,
  name: '',
};

export const createPortfolioReducer = (
  state: CreatingPortfolio = initialState,
  action: CreatePortfolioAction
): CreatingPortfolio => {
  switch (action.type) {
    case ActionType.CreateNewPortfolio:
      return { ...state, name: action.name, amount: action.amount };
    default:
      return state;
  }
};
