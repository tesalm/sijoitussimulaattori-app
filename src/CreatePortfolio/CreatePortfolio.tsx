import React from 'react';
import { Button, Text, TextInput, View } from 'react-native';
import { NavigationScreenProps } from 'react-navigation';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';
import validator from 'validator';

import { t } from '../assets/i18n';
import { RouteName } from '../navigation/routes';
import { RootState } from '../redux/reducers';
import { CreateNewPortfolio } from './actions';
import { CreatePortfolioStyles as styles } from './styles';

export interface CreatePortfolioProps {
  onCreatePortfolio: typeof CreateNewPortfolio;
}

type CreatePortfolioPropsWithNavigation = CreatePortfolioProps &
  NavigationScreenProps;

interface CreatePortfolioState {
  inputNumber: string;
  amount: number;
  name: string;
  nameError: boolean;
  amountError: boolean;
  errorMessageTranslationKey: string;
}

class CreatePortfolio extends React.Component<
  CreatePortfolioPropsWithNavigation,
  CreatePortfolioState
> {
  constructor(props: CreatePortfolioPropsWithNavigation) {
    super(props);
    this.state = {
      inputNumber: '',
      amount: NaN,
      name: '',
      nameError: true,
      amountError: true,
      errorMessageTranslationKey: '',
    };
  }

  sanitize(input: string) {
    const trimWhitespaces = validator.trim(input);
    if (trimWhitespaces === '') {
      this.setState({
        errorMessageTranslationKey: 'InputErrors.NameEmpty',
      });
    } else {
      this.setState({ name: input, nameError: false });
    }
  }

  validateAmount(input: string) {
    const trimWhitespaces = validator.trim(input);
    const commaToPoint: string = trimWhitespaces.replace(/,/g, '.');
    if (commaToPoint === '') {
      this.setState({
        errorMessageTranslationKey: 'InputErrors.NotNumber',
      });
    } else if (validator.isNumeric(commaToPoint)) {
      this.setState({
        amount: validator.toFloat(commaToPoint),
        amountError: false,
      });
    } else {
      this.setState({
        errorMessageTranslationKey: 'InputErrors.NotNumber',
      });
    }
  }

  validate() {
    const { onCreatePortfolio } = this.props;
    this.sanitize(this.state.name);
    this.validateAmount(this.state.inputNumber);
    if (!this.state.nameError && !this.state.amountError) {
      onCreatePortfolio(this.state.name, this.state.amount);
      this.props.navigation.navigate(RouteName.Home);
    }
  }

  showErrors() {
    return this.state.nameError || this.state.amountError ? (
      <Text style={styles.error}>
        {t(this.state.errorMessageTranslationKey)}
      </Text>
    ) : null;
  }

  static navigationOptions = { title: t('CreatePortfolio.Title') };
  render() {
    return (
      <View style={styles.container}>
        <Text style={styles.main}>
          {t('CreatePortfolio.PortfolioNameInputTitle')}
        </Text>
        <TextInput
          style={styles.textinput}
          ref="name"
          onChangeText={(name) => this.setState({ name })}
          value={this.state.name}
        />
        <Text style={styles.main}>
          {t('CreatePortfolio.PortfolioAmountInputTitle')}
        </Text>
        <TextInput
          style={styles.textinput}
          keyboardType="numeric"
          ref="amount"
          onChangeText={(inputNumber) => this.setState({ inputNumber })}
          value={this.state.inputNumber}
        />
        <Button
          color="#004d40"
          title={t('CreatePortfolio.ConfirmButtonText')}
          onPress={() => this.validate()}
        />
        <View>{this.showErrors()}</View>
      </View>
    );
  }
}

const mapStateToProps = (state: RootState) => ({
  counterValue: state.counter.counterValue,
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      onCreatePortfolio: CreateNewPortfolio,
    },
    dispatch
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CreatePortfolio);

export { CreatePortfolio as CreatePortfolioTest };
