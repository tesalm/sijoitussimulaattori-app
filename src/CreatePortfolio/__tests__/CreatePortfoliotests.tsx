import * as React from 'react';
import renderer from 'react-test-renderer';

import { CreatePortfolioProps, CreatePortfolioTest } from '../CreatePortfolio';

describe('Create Portfolio', () => {
  //Mock for navigation.
  const navigationMock: any = {};

  const defaultCreatePortfolioProps: CreatePortfolioProps = {
    onCreatePortfolio: jest.fn(),
  };

  it('renders correctly', async () => {
    const component = renderer
      .create(
        <CreatePortfolioTest
          {...defaultCreatePortfolioProps}
          {...navigationMock}
        />
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  //TODO: Testing text inputs
  it('input test #1', () => {
    const component = renderer
      .create(
        <CreatePortfolioTest
          {...defaultCreatePortfolioProps}
          {...navigationMock}
        />
      )
      .getInstance();
    const nameAndAmount = { name: '', amount: 10 };
  });
});
