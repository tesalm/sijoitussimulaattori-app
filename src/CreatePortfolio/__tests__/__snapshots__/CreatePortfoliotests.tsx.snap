// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Portfolio renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "stretch",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "flex-start",
      "marginBottom": 8,
      "marginLeft": 16,
      "marginRight": 16,
      "marginTop": 8,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#004D40",
        "fontFamily": "Roboto",
        "fontSize": 16,
        "fontWeight": "bold",
        "textAlign": "left",
      }
    }
  >
    Name of the portfolio
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    style={
      Object {
        "borderBottomColor": "#000000",
        "borderBottomWidth": 1,
        "color": "#e8a30c",
        "height": 40,
        "marginBottom": 8,
        "marginTop": 8,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Text
    style={
      Object {
        "color": "#004D40",
        "fontFamily": "Roboto",
        "fontSize": 16,
        "fontWeight": "bold",
        "textAlign": "left",
      }
    }
  >
    Initial amount of money
  </Text>
  <TextInput
    allowFontScaling={true}
    keyboardType="numeric"
    onChangeText={[Function]}
    style={
      Object {
        "borderBottomColor": "#000000",
        "borderBottomWidth": 1,
        "color": "#e8a30c",
        "height": 40,
        "marginBottom": 8,
        "marginTop": 8,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
            Object {
              "color": "#004d40",
            },
          ]
        }
      >
        Create Portfolio
      </Text>
    </View>
  </View>
  <View>
    <Text
      style={
        Object {
          "color": "#ff0000",
        }
      }
    >
      
    </Text>
  </View>
</View>
`;
